null element in aws_ecs_task_definition container_definitions causes provider panic

In creating a list of json encoded container defintions to pass to a ecs task definition I introduced a `null` default value for an optional side-car container_definition in the task. If that null is not filtered out the aws provider has a panic during terraform plan/apply. This was a user error on my part as I should have run a `compact` on the list but also seems like the provider should better handle possibility of null/empty inputs here.

<!--- Please keep this note for the community --->

### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

<!--- Thank you for keeping this note for the community --->

### Terraform CLI and Terraform AWS Provider Version

<!--- Please run `terraform -v` to show the Terraform core version and provider version(s). If you are not running the latest version of Terraform or the provider, please upgrade because your issue may have already been fixed. [Terraform documentation on provider versioning](https://www.terraform.io/docs/configuration/providers.html#provider-versions). --->
Terraform v0.15.3
on darwin_amd64
+ provider registry.terraform.io/hashicorp/aws v3.42.0
+ provider registry.terraform.io/hashicorp/external v2.1.0
+ provider registry.terraform.io/hashicorp/local v2.1.0

### Affected Resource(s)

<!--- Please list the affected resources and data sources. --->

* aws_ecs_task_definition

### Terraform Configuration Files

<!--- Information about code formatting: https://help.github.com/articles/basic-writing-and-formatting-syntax/#quoting-code --->

Please include all Terraform configurations required to reproduce the bug. Bug reports without a functional reproduction may be closed without investigation.

```hcl
resource "aws_ecs_task_definition" "this" {
  container_definitions = jsonencode(flatten([
    var.json_map_container_definitions,
    [module.fluentbit_firelens_container_definition.json_map_object]
  ]))

  cpu                                   = var.task_cpu
  family                               = local.namespace
  memory                           = var.task_memory
  network_mode                = "awsvpc"
  tags                                  = local.tags
  requires_compatibilities = ["FARGATE"]
  execution_role_arn         = aws_iam_role.execution.arn
  task_role_arn                  = aws_iam_role.task.arn
}
```
In the snippet above the `var.json_map_container_definitions` is a list of maps produced by the cloudposse container-definition module as a convenience method to create ECS container definition json and includes 1 null item. The parent module has an optional sidecar pre-defined, but sets the container definition object to `null` if not selected for use. That null persists into the jsonencoding and the provider panics when it tries to access the task environment variables for sorting.

### Debug Output

<!---
Please provide a link to a GitHub Gist containing the complete debug output. Please do NOT paste the debug output in the issue; just paste a link to the Gist.

To obtain the debug output, see the [Terraform documentation on debugging](https://www.terraform.io/docs/internals/debugging.html).
--->

### Panic Output

https://gist.github.com/zswanson/d6e9058e17151066f1fe2de1384abcbb

### Expected Behavior

Output indicating that a null element was found during the plan and terraform to gracefully terminate the plan

### Actual Behavior

Terraform provider panics, plan/apply fails 

### Steps to Reproduce

<!--- Please list the steps required to reproduce the issue. --->

1. `terraform plan`

### Important Factoids

<!--- Are there anything atypical about your accounts that we should know? For example: Running in EC2 Classic? --->

### References

<!---
Information about referencing Github Issues: https://help.github.com/articles/basic-writing-and-formatting-syntax/#referencing-issues-and-pull-requests

Are there any other GitHub issues (open or closed) or pull requests that should be linked here? Vendor documentation? For example:
--->

* #0000

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0xa8 pc=0x516c61b]

goroutine 252 [running]:
github.com/terraform-providers/terraform-provider-aws/aws.containerDefinitions.OrderEnvironmentVariables(0xc001d9e5a0, 0x4, 0x4)
	/opt/teamcity-agent/work/5d79fe75d4460a2f/src/github.com/hashicorp/terraform-provider-aws/aws/ecs_task_definition_equivalency.go:107 +0x3b
github.com/terraform-providers/terraform-provider-aws/aws.resourceAwsEcsTaskDefinition.func2(0x6387780, 0xc001d8f0b0, 0x15, 0x0)
	/opt/teamcity-agent/work/5d79fe75d4460a2f/src/github.com/hashicorp/terraform-provider-aws/aws/resource_aws_ecs_task_definition.go:83 +0x85
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.schemaMap.diffString(0xc000b061b0, 0x7a34099, 0x15, 0xc000b09180, 0xc001c35508, 0x8519d70, 0xc001781480, 0x203000, 0x0, 0x0)
	/opt/teamcity-agent/work/5d79fe75d4460a2f/pkg/mod/github.com/hashicorp/terraform-plugin-sdk/v2@v2.6.1/helper/schema/schema.go:1375 +0x6e3
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.schemaMap.diff(0xc000b061b0, 0x7a34099, 0x15, 0xc000b09180, 0xc001d9e360, 0x8519d70, 0xc001781480, 0xc001a2b900, 0x0, 0x0)
	/opt/teamcity-agent/work/5d79fe75d4460a2f/pkg/mod/github.com/hashicorp/terraform-plugin-sdk/v2@v2.6.1/helper/schema/schema.go:957 +0x4e8
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.schemaMap.Diff(0xc000b061b0, 0x84ff548, 0xc001a50740, 0xc000db8310, 0xc001a7fcb0, 0x7b4b1d0, 0x6c6f040, 0xc001052600, 0xbb4c700, 0x6387780, ...)
	/opt/teamcity-agent/work/5d79fe75d4460a2f/pkg/mod/github.com/hashicorp/terraform-plugin-sdk/v2@v2.6.1/helper/schema/schema.go:522 +0x219
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*Resource).SimpleDiff(0xc000a71dc0, 0x84ff548, 0xc001a50740, 0xc000db8310, 0xc001a7fcb0, 0x6c6f040, 0xc001052600, 0x0, 0x0, 0x0)
	/opt/teamcity-agent/work/5d79fe75d4460a2f/pkg/mod/github.com/hashicorp/terraform-plugin-sdk/v2@v2.6.1/helper/schema/resource.go:506 +0xa5
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*GRPCProviderServer).PlanResourceChange(0xc0007a05b8, 0x84ff548, 0xc001a50740, 0xc0019e1450, 0xc001a50740, 0x77ee4a0, 0xc001a7e900)
	/opt/teamcity-agent/work/5d79fe75d4460a2f/pkg/mod/github.com/hashicorp/terraform-plugin-sdk/v2@v2.6.1/helper/schema/grpc_provider.go:693 +0x7c5
github.com/hashicorp/terraform-plugin-go/tfprotov5/server.(*server).PlanResourceChange(0xc002014600, 0x84ff5f0, 0xc001a50740, 0xc000db81c0, 0xc002014600, 0xc001a7e9c0, 0xc001ff1ba0)
	/opt/teamcity-agent/work/5d79fe75d4460a2f/pkg/mod/github.com/hashicorp/terraform-plugin-go@v0.3.0/tfprotov5/server/server.go:315 +0xb5
github.com/hashicorp/terraform-plugin-go/tfprotov5/internal/tfplugin5._Provider_PlanResourceChange_Handler(0x77ee4a0, 0xc002014600, 0x84ff5f0, 0xc001a7e9c0, 0xc00128ac60, 0x0, 0x84ff5f0, 0xc001a7e9c0, 0xc001a3e600, 0x10a3)
	/opt/teamcity-agent/work/5d79fe75d4460a2f/pkg/mod/github.com/hashicorp/terraform-plugin-go@v0.3.0/tfprotov5/internal/tfplugin5/tfplugin5_grpc.pb.go:362 +0x214
google.golang.org/grpc.(*Server).processUnaryRPC(0xc0009a5340, 0x85203f8, 0xc00035a600, 0xc0019d6d00, 0xc0013bbec0, 0xbb0a4a8, 0x0, 0x0, 0x0)
	/opt/teamcity-agent/work/5d79fe75d4460a2f/pkg/mod/google.golang.org/grpc@v1.32.0/server.go:1194 +0x52b
google.golang.org/grpc.(*Server).handleStream(0xc0009a5340, 0x85203f8, 0xc00035a600, 0xc0019d6d00, 0x0)
	/opt/teamcity-agent/work/5d79fe75d4460a2f/pkg/mod/google.golang.org/grpc@v1.32.0/server.go:1517 +0xd0c
google.golang.org/grpc.(*Server).serveStreams.func1.2(0xc000efeb40, 0xc0009a5340, 0x85203f8, 0xc00035a600, 0xc0019d6d00)
	/opt/teamcity-agent/work/5d79fe75d4460a2f/pkg/mod/google.golang.org/grpc@v1.32.0/server.go:859 +0xab
created by google.golang.org/grpc.(*Server).serveStreams.func1
	/opt/teamcity-agent/work/5d79fe75d4460a2f/pkg/mod/google.golang.org/grpc@v1.32.0/server.go:857 +0x1fd
```

https://github.com/hashicorp/terraform-provider-aws/blob/d7e16d776d9f9e13945513caf46aa3cab8578442/aws/ecs_task_definition_equivalency.go#L105-L111

--- a/.changelog/27263.txt
+++ b/.changelog/27263.txt
@@ -1,0 +1,3 @@
+```release-note:bug
+resource/aws_ecs_task_definition: prevent panic when supplying a 'null' container definition
+```

--- a/.changelog/27263.txt
+++ b/.changelog/27263.txt
@@ -1,3 +1,3 @@
 ```release-note:bug
-resource/aws_ecs_task_definition: prevent panic when supplying a 'null' container definition
+resource/aws_ecs_task_definition: Prevent panic when supplying a `null` value in `container_definitions`
 ```

--- a/internal/service/ecs/task_definition.go
+++ b/internal/service/ecs/task_definition.go
@@ -1208,6 +1208,12 @@ func expandContainerDefinitions(rawDefinitions string) ([]*ecs.ContainerDefiniti
 		return nil, fmt.Errorf("Error decoding JSON: %s", err)
 	}
 
+	for i, c := range definitions {
+		if c == nil {
+			return nil, fmt.Errorf("invalid container definition supplied at index (%d)", i)
+		}
+	}
+
 	return definitions, nil
 }
 

--- a/internal/service/ecs/task_definition_test.go
+++ b/internal/service/ecs/task_definition_test.go
@@ -1002,6 +1002,23 @@ func TestAccECSTaskDefinition_inferenceAccelerator(t *testing.T) {
 	})
 }
 
+func TestAccECSTaskDefinition_invalidContainerDefinition(t *testing.T) {
+	rName := sdkacctest.RandomWithPrefix(acctest.ResourcePrefix)
+
+	resource.ParallelTest(t, resource.TestCase{
+		PreCheck:                 func() { acctest.PreCheck(t) },
+		ErrorCheck:               acctest.ErrorCheck(t, ecs.EndpointsID),
+		ProtoV5ProviderFactories: acctest.ProtoV5ProviderFactories,
+		CheckDestroy:             testAccCheckTaskDefinitionDestroy,
+		Steps: []resource.TestStep{
+			{
+				Config:      testAccTaskDefinitionConfig_invalidContainerDefinition(rName),
+				ExpectError: regexp.MustCompile(`invalid container definition supplied at index \(1\)`),
+			},
+		},
+	})
+}
+
 func testAccTaskDefinitionConfig_proxyConfiguration(rName string, containerName string, proxyType string,
 	ignoredUid string, ignoredGid string, appPorts string, proxyIngressPort string, proxyEgressPort string,
 	egressIgnoredPorts string, egressIgnoredIPs string) string {
@@ -2615,4 +2632,33 @@ resource "aws_fsx_windows_file_system" "test" {
   throughput_capacity = 8
 }
 `)
+}
+
+func testAccTaskDefinitionConfig_invalidContainerDefinition(rName string) string {
+	return fmt.Sprintf(`
+resource "aws_ecs_task_definition" "test" {
+  family = %[1]q
+
+  container_definitions = <<TASK_DEFINITION
+[
+	{
+		"cpu": 10,
+		"command": ["sleep", "10"],
+		"entryPoint": ["/"],
+		"essential": true,
+		"image": "mongodb",
+		"memory": 128,
+		"name": "mongodb",
+		"portMappings": [
+			{
+				"containerPort": 28017,
+				"hostPort": 28017
+			}
+		]
+	},
+	null
+]
+TASK_DEFINITION
+}
+`, rName)
 }

